src/libprojectM/Renderer/MilkdropNoise.cpp

```diff
@@ -1,42 +1,131 @@
 #include "MilkdropNoise.hpp"
 
+#include "projectM-opengl.h"
+
 #include <chrono>
 #include <random>
 
+// Missing in macOS SDK. Query will most certainly fail, but then use the default format.
+#ifndef GL_TEXTURE_IMAGE_FORMAT
+#define GL_TEXTURE_IMAGE_FORMAT 0x828F
+#endif
+
 namespace libprojectM {
 namespace Renderer {
 
-MilkdropNoise::MilkdropNoise()
+auto MilkdropNoise::LowQuality() -> std::shared_ptr<Texture>
+{
+    GLuint texture{};
+    {
+        auto textureData = generate2D(256, 1);
+
+        glGenTextures(1, &texture);
```

Comment:
This allocates a new texture each time this is called. That's fine if each of these is only called once by TextureManager, but if it's called twice it'll generate a new texture with different values. No action needed, just a note. In theory you could initialize a static field in this class and return it if it's been constructed, but I don't think the change is worth it since this should only be called once.

Reply:
I don't understand the issue at hand here.
The noise is pseudo-random, so for each call I'd actually _expect_ it to be different (or, rather even require it). Plus, this function is actually there to create a new texture upon each call (not reuse any existing one), because users might want different noise textures for each preset (currently, these are just generated once per projectM instance).

Reply:
The previous implementation defined texture data and initialized it, and that texture data is referenced directly. The new implementation *creates* texture data each time it's called. It doesn't matter as long as this isn't called multiple times for a single preset (which it isn't) so it doesn't matter. No changes required.

Reply:
The series of events was changed, but technically, it does the exact same thing as before. Just to compare, here's how it worked previously each time `TextureManager::Preload()` was called:

1. Create an instance of MilkdropNoise.
2. The MilkdropNoise constructor called each static noise generator function once and stored the resulting texture data as a raw pointer
3. For each noise texture image, a GL texture was then created and the image data _copied_ into it (via `glTexture2D`/`glTexture3D`).
4. A projectM `Texture` instance was then created with the resulting GL texture ID for use in presets, stored as a shared pointer.
5. The MilkdropNoise instance was deleted at the end of the `Preload` function, including the image data.

Now the MilkdropNoise class is purely static, and texture creation was just moved into each noise function, like this:

1. For each noise texture, the static generator function in Milkdrop noise is called, which then does:
  1. Generate the pixel data for the requested noise image.
  2. Create a GL texture and upload the image data into it.
  3. Delete the local pixel data.
  4. Create a `Texture` instance with the just-created texture ID and return it as a shared pointer.
2. In `Preload()`, the returned shared_ptr instance is then added to the texture storage.